social_django/views.py

```diff
@@ -39,21 +43,86 @@ def disconnect(request, backend, association_id=None):
+            # from the provider, with no set maximum. Use that.
+            session_expiry = received_expiration_time
+        else:
+            # We received an expiration time from the backend, and we also
+            # have a set maximum session length. Use the smaller of the two.
+            session_expiry = min(received_expiration_time, max_session_length)
+    else:
+        # If there's an explicitly-set maximum session length, use that
+        # even if we don't want to retrieve session expiry times from
+        # the backend. If there isn't, then use the platform default.
+        if max_session_length is None:
+            session_expiry = DEFAULT_SESSION_TIMEOUT
+        else:
+            session_expiry = max_session_length
+
+    return session_expiry
+
+
 def _do_login(backend, user, social_user):
     user.backend = '{0}.{1}'.format(backend.__module__,
                                   backend.__class__.__name__)
+    # Get these details early to avoid any issues involved in the
+    # session switch that happens when we call login().
+    enable_session_expiration = backend.setting('SESSION_EXPIRATION', False)
+    max_session_length_setting = backend.setting('MAX_SESSION_LENGTH', None)
+
+    # Log the user in, creating a new session.
     login(backend.strategy.request, user)
-    if backend.setting('SESSION_EXPIRATION', False):
-        # Set session expiration date if present and enabled
-        # by setting. Use last social-auth instance for current
-        # provider, users can associate several accounts with
-        # a same provider.
-        expiration = social_user.expiration_datetime()
-        if expiration:
-            try:
-                backend.strategy.request.session.set_expiry(
-                    expiration.seconds + expiration.days * 86400
-                )
-            except OverflowError:
-                # Handle django time zone overflow
-                backend.strategy.request.session.set_expiry(None)
+
+    # Make sure that the max_session_length value is either an integer or
+    # None. Because we get this as a setting from the backend, it can be set
+    # to whatever the backend creator wants; we want to be resilient against
+    # unexpected types being presented to us.
+    try:
+        max_session_length = int(max_session_length_setting)
+    except (TypeError, ValueError):
+        # We got a response that doesn't look like a number; use the default.
+        max_session_length = None
+
+    # Get the session expiration length based on the maximum session length
+    # setting, combined with any session length received from the backend.
+    session_expiry = get_session_timeout(
+        social_user,
+        enable_session_expiration=enable_session_expiration,
+        max_session_length=max_session_length,
+    )
+
+    try:
+        # Set the session length to our previously determined expiry length.
+        backend.strategy.request.session.set_expiry(session_expiry)
```

Comment:
If the session's expiration was set in the pipeline by calling `strategy.request.session.set_expiry()`, the expiration will be overwritten by this line. Is there a way to prevent this from happening?